### PR TITLE
Aho-Corasick multiple string matching algorithm

### PR DESCRIPTION
**Pull request by pmikolajczyk41**
Brief implementation of https://cr.yp.to/bib/1975/aho.pdf with tests and a short description.

There were some minor problems with Pylint:
```python
# pylint: disable=too-few-public-methods
class AhoCorasickAutomaton:
  ...
```
i.e. I've encapsulated the creation of the automaton within a class, but eventually it exposes only one public method (for searching). I could either disable Pylint's screaming or add some dummy method like ```__str__()```.

Moreover in the test script, Pylint has a problem with:
```python
for p in patterns:
  ...
  indices = map(lambda i: (i, i + len(p)), starts)
```
It is an open issue raised here: https://github.com/PyCQA/pylint/issues/3107

### CODE REVIEW

**Comment by rafalalgo** [comment]
Probably you can change
`indices = map(lambda i: (i, i + len(p)), starts)` 
into
`indices = [(i, i + len(p)) for i in starts]`

then pylint will be happy :smile:

**Comment by pmikolajczyk41** [comment]
Thank you, of course it could be easily fixed -- I must have had some kinda blackout :)

File: text/PiotrMikolajczyk/main.tex [hunk 1]
```diff
@@ -0,0 +1,249 @@
+\documentclass{article}
```
[the remark this answers is not in the document]
Comment (krzysztof-turowski) [reply text/PiotrMikolajczyk/main.tex:1]: OK, I did miss it.

File: exact_multiple_string_matching/aho_corasick.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+          fallback = current.fail()
+          current.update_nxt(a, fallback.nxt(a))
+      current.use_only_nxt()
+
+  def find_occurrences(self, text, n):
+    state = self._root
+    for i in range(1, n + 1):
+      state = state.nxt(text[i])
+      for keyword_len in state.output():
+        yield i - keyword_len + 1, i + 1
```
[the remark this answers is not in the document]
Comment (pmikolajczyk41) [reply exact_multiple_string_matching/aho_corasick.py:89]: Firstly, I implemented the 2. proposition. After realizing, that it is possible to completely abandon storing keywords in the automaton, I think that the current solution is much more sparing and we do not lose any clarity. The only reasonable situation I can imagine when we would need particular pattern is when encountering it, but then we can just take a corresponding substring from the text.
Surely, I'd be glad to hear others' opinion and, if needed, change the realization.

File: text/PiotrMikolajczyk/main.tex [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+  def __init__(self, keywords, alphabet):
+    self._root = AhoCorasickAutomaton.Node()
+    self._construct_goto(keywords, alphabet)
+    self._construct_fail(alphabet)
+    self._construct_next(alphabet)
+\end{minted}
+
+Bardzo łatwo zauważyć, że każdą z nich wykonać można w czasie $\Theta(\sum_{i\in[n]}|k_i|)$ -- za każdym razem przechodzimy po całym drzewie $\trie$ odwiedzając każdy wierzchołek dokładnie raz oraz wykonując $\Theta(1)$ operacji w każdym z nich. Zakładamy również, że łączenie zbiorów $\out$ wykonuje się w czasie stałym (jak np. listy wiązane).
+
+Pozostaje zastanowić się, ile czasu zajmie przeczytanie tekstu $T$ ($|T|=m$) i wypisanie wszystkich wystąpień wzorców. Moc $\Sigma$ traktujemy jak stałą.
```
[the remark this answers is not in the document]
Comment (pmikolajczyk41) [reply text/PiotrMikolajczyk/main.tex:224]: I agree, as the size of the automaton is bounded by the product of the sum of keywords' length and alphabet cardinality. Every phase -- either construction or searching -- is linear with respect to the size of the graph

File: text/PiotrMikolajczyk/main.tex [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+    można również spojrzeć na to w ten sposób, że przechodząc automatem $\A$ wzdłuż $T$, pamiętamy dwa wskaźniki w $T$ -- słowo pomiędzy nimi to właśnie to, które jest generowane przez ścieżkę z korzenia do obecnego wierzchołka; przejście $\goto$ przesuwa 'prawy' wskaźnik o 1 w prawo, przejście $\fail$ przesuwa 'lewy' wskaźnik o dodatnią liczbę pozycji w prawą, ale na tyle małą by nie przeskoczyć prawego;
+    
+    \item używając funkcji $\next$, przy każdym przeczytanym symbolu z $T$ wykonujemy zawsze jedno przejście -- $m$ ruchów
+\end{itemize}
+
+Problem może się pojawić, gdy będziemy chcieli wypisywać wszystkie wystąpienia wzorców. Pesymistycznym przykładem jest $\K=\{a,a^2,a^3,...,a^n\}$ oraz $T=a^m$. Liczba wystąpień będzie $\Theta(nm)$. Jednakże, nie jest to wada zastosowanego algorytmu -- każdy inny algorytm zgłaszający wszystkie wystąpienia wzorców musiałby wykonać co najmniej tyle samo pracy. 
+
+Zatem możemy określić złożoność czasową wyszukiwania wielu wzorców w tekście za pomocą automatu Aho-Corasick jako liniową w stosunku do sumy długości wszystkich wzorców oraz ilości ich wystąpień.
+
+\section{Zastosowanie do wzorców 2D}
```
[the remark this answers is not in the document]
Comment (pmikolajczyk41) [reply text/PiotrMikolajczyk/main.tex:239]: Hoped you'll like it :) This application was presented as a curiosity during ADS lecture

File: exact_multiple_string_matching/aho_corasick.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (pmikolajczyk41) [reply exact_multiple_string_matching/aho_corasick.py:89]: The first option is the one I would rather not to choose -- if there is any occurrence we will return _something_ anyway. In any other situation we will lose information, providing which does not cost much.

File: text/PiotrMikolajczyk/main.tex [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+\begin{minted}[xleftmargin=20pt,linenos]{python}
+class AhoCorasickAutomaton:
+  def __init__(self, keywords, alphabet):
+    self._root = AhoCorasickAutomaton.Node()
+    self._construct_goto(keywords, alphabet)
+    self._construct_fail(alphabet)
+    self._construct_next(alphabet)
+\end{minted}
+
+Bardzo łatwo zauważyć, że każdą z nich wykonać można w czasie $\Theta(\sum_{i\in[n]}|k_i|)$ -- za każdym razem przechodzimy po całym drzewie $\trie$ odwiedzając każdy wierzchołek dokładnie raz oraz wykonując $\Theta(1)$ operacji w każdym z nich. Zakładamy również, że łączenie zbiorów $\out$ wykonuje się w czasie stałym (jak np. listy wiązane).
```
[the remark this answers is not in the document]
Comment (pmikolajczyk41) [reply text/PiotrMikolajczyk/main.tex:222]: For now I will use $k$ for the number of patters, but maybe I will not take care about the longest pattern -- I will leave the sum of lengths as it is, especially that $m$ is currently the $T$'s length :)
But of course, if you wish, I can still exchange sum for the length upperbound

**Comment by pmikolajczyk41** [comment]
I have applied almost all suggestions -- please review them and let me know what else should be done

File: exact_multiple_string_matching/aho_corasick.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (krzysztof-turowski) [reply exact_multiple_string_matching/aho_corasick.py:89]: Sure, I agree that you may completely abandon storing keywords in the automaton - however I was thinking in terms of clarity of explanation vs. optimality.

As for the explanations: I would think that it's possible that we would like for a list of patterns w_1, w_2, ..., w_k to build an automaton and answer queries with arrays W[1..k] storing how many times each string appeared without any further comparisons. Equivalently, we could yield pairs (i, pos), i.e. store i's at the respective nodes. But I understand your concerns, but considering the above I guess returning (w_i, pos) maybe with a short explanation that (i, pos) can be done very similar is more readable by laymen. At least it's just my impression.

File: text/PiotrMikolajczyk/main.tex (hunk 5 above)
[the remark this answers is not in the document]
Comment (krzysztof-turowski) [reply text/PiotrMikolajczyk/main.tex]: Well, I used through my notes $|t| = n$, $|w| = m$ (or $\max_i |w_i| = m$ if there were multiple), but OK, this I can correct during my merge run.

File: exact_multiple_string_matching/aho_corasick.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (pmikolajczyk41) [reply exact_multiple_string_matching/aho_corasick.py:89]: It's hard to disagree :) I think that in terms of this repo at least, consistency and clarity are priorities, so maybe let's wait on @rafalalgo 's point of view and I will adapt properly

File: text/PiotrMikolajczyk/main.tex (hunk 5 above)
[the remark this answers is not in the document]
Comment (pmikolajczyk41) [reply text/PiotrMikolajczyk/main.tex]: I was convinced after realizing that I have breached the convention also for $t$ -- my apologies. The patch is on its way

File: exact_multiple_string_matching/aho_corasick.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (rafalalgo) [reply exact_multiple_string_matching/aho_corasick.py:89]: In Commentz-Walter algorithm implementation, my proposition is that for a list of patterns w_1, w_2, ..., w_k to build an automaton and then to answer queries, we could yield pairs (pattern, pos), i.e. [ ('acb', 1), ('aba', 5), ('aba', 7), ('aba', 9), ('aba', 11) ]. It is a proposition from original Commentz-Walter paper :)

[the remark this answers is not in the document]
Comment (pmikolajczyk41) [reply exact_multiple_string_matching/aho_corasick.py:89]: Resolved :)

**Pull request merged**